cores/arduino/soc_i2c.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * soc_i2c.c - i2c library layer
+ *
+ * Copyright (C) 2015 Intel Corporation
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include <stdbool.h>
+#include "soc_i2c.h"
+#include "i2c.h"
+#include "variant.h"
+
+#define TIMEOUT_MS 16
+
+static volatile uint8_t soc_i2c_tx_complete;
+static volatile uint8_t soc_i2c_rx_complete;
+static volatile uint8_t soc_i2c_err_detect;
+static volatile uint32_t soc_i2c_err_source;
+
+static volatile uint8_t soc_i2c_slave = 0;
+
+static void soc_i2c_rx(uint32_t dev_id)
+{
+	soc_i2c_rx_complete = 1;
+}
+
+static void soc_i2c_tx(uint32_t dev_id)
+{
+	soc_i2c_tx_complete = 1;
+}
+
+static void soc_i2c_err(uint32_t dev_id)
+{
+	soc_i2c_err_detect = 1;
+	soc_i2c_err_source = dev_id;
+}
+
+static int soc_i2c_wait_rx_or_err(bool no_stop){
+	uint64_t timeout = TIMEOUT_MS;
+	while(timeout--) {
+		if (soc_i2c_err_detect) {
+			if (soc_i2c_err_source & I2C_ABRT_7B_ADDR_NOACK )
+			{
+				return I2C_ERROR_ADDRESS_NOACK; // NACK on transmit of address
+			}
+			else if (soc_i2c_err_source & I2C_ABRT_TXDATA_NOACK )
+			{
+				return I2C_ERROR_DATA_NOACK; // NACK on transmit of data
+			}
+			else
+			{
+				return I2C_ERROR_OTHER; // other error
+			}
+		}
+		if (!no_stop) {
+			if (soc_i2c_rx_complete) {
+				return I2C_OK;
+			}
+		}
+		delay(1);
+	}
+	if (!no_stop)
+		return I2C_TIMEOUT;
+	else
+		return I2C_OK;
+}
+
+static int soc_i2c_wait_tx_or_err(bool no_stop){
+	uint64_t timeout = TIMEOUT_MS;
+	while(timeout--) {
+		if (soc_i2c_err_detect) {
+			if (soc_i2c_err_source & I2C_ABRT_7B_ADDR_NOACK )
+			{
+				return I2C_ERROR_ADDRESS_NOACK; // NACK on transmit of address
+			}
+			else if (soc_i2c_err_source & I2C_ABRT_TXDATA_NOACK )
+			{
+				return I2C_ERROR_DATA_NOACK; // NACK on transmit of data
+			}
+			else
+			{
+				return I2C_ERROR_OTHER; // other error
+			}
+		}
+		if (!no_stop) {
+			if (soc_i2c_tx_complete) {
+			return I2C_OK;
+			}
+		}
+		delay(1);
```

Comment:
I see a lot of loops that have been "spaced out" using delay(1). That's fine, it shouldn't have any negative effects since I2C is quite slow anyway, but I don't think it will gain anything either. Remember, this is for a single-threaded MCU device.